pkg/reconciler/controllers/remediator.go

```diff
@@ -0,0 +1,348 @@
+// and remediate the object to match the declared resource.
+type Remediator struct {
+	// ControllerCtx is set in reconciler/main.go and is canceled by the Finalizer.
+	ControllerCtx context.Context
+	// DoneCh indicates the Remediator stops processing new requests.
+	// It is a signal for the Finalizer to continue to finalize managed resources.
+	DoneCh chan struct{}
+	// SyncScope is the scope of the reconciler, either a namespace or ':root'
+	SyncScope declared.Scope
+	// SyncName is the name of the RootSync or RepoSync.
+	SyncName string
+	// Applier is the client to correct drift.
+	syncerreconcile.Applier
+	// declared.Resources is the cache of the declared resource list. It is used
+	// as the source of truth for drift correction.
+	*declared.Resources
+	// cache.RemediateResources is the cache of the remediate resource list. It is
+	// used to get the actual Kubernetes object for a particular GKNN.
+	*cache.RemediateResources
+	// Remediator is the actor to check whether it is ready to start reconciliation.
+	*remediator.Remediator
+}
+
+// Reconcile processes the remediate request to correct drift.
+func (r *Remediator) Reconcile(_ context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if errors.Is(r.ControllerCtx.Err(), context.Canceled) {
+		// Drain the request queue
+		klog.V(3).Infof("Stop remediating new objects as the context is canceled..")
+		return ctrl.Result{}, nil
+	}
+
+	klog.V(1).Infof("New remediator reconciliation for %q", req.Name)
+
+	id, err := core.FromID(req.Name)
+	if err != nil {
+		klog.Errorf("invalid request: %v", err)
+		return ctrl.Result{}, nil
+	}
+
+	// marks the Remediator is working to block the Applier if it can remediate.
+	if !r.StartRemediating() {
+		// Mark the object as done so that it can be requeued later when the Applier is not running.
+		klog.V(1).Infof("Skip remediating object %s because the Remediator is paused", id)
+		r.DoneObj(id)
+		return ctrl.Result{}, nil
+	}
+	// Mark the Remediator as done before it exits, so the Syncer can start the apply work.
+	defer r.DoneRemediating()
+
+	obj, found := r.GetObjs()[id]
+	if !found {
+		klog.V(4).Infof("Skip remediating object %q because it is no longer on the remediate list", id)
+		return ctrl.Result{}, nil
+	}
+
+	// Get the ResourceVersion of the object from the last remediation.
+	lastRV := r.LastRemediateRV(id)
+	// Get the current ResourceVersion.
+	currentRV := obj.GetResourceVersion()
+
+	var toRemediate client.Object
+	if cache.WasDeleted(r.ControllerCtx, obj) {
+		// Passing a nil Object to the reconciler signals that the accompanying ID
+		// is for an Object that was deleted.
+		toRemediate = nil
+		// If the object was deleted, it has no ResourceVersion, use the constant.
+		currentRV = cache.DeletedResourceVersion
+	} else {
+		toRemediate = obj
+	}
+
+	// If the current ResourceVersion was last remediated, skip remediation because
+	// it should match the declared resource.
```

Comment:
Should we also skip remediation when the only change on the obj is its status field? ResourceVersion changes on status updates, too.

Reply:
This might not be easily done given that comparing `generation` won't capture metadata updates, so can be done later as an enhancement.